# Anomaly Detection in Azure ML Studio

Example anomaly detection machine learning job using a hybrid of Azure AI Machine Learning console resources combined with native Python in this Jupyter notebook which is intended to be uploaded and executed from 'Authoring -> Notebooks' in the web console. 

### Requirements for implementing the Proof of Concept

1. You should already have created an Azure account, and created a [Subscription](https://techcommunity.microsoft.com/discussions/azure/understanding-azure-account-subscription-and-directory-/34800) and a [Workspace](https://learn.microsoft.com/en-us/azure/machine-learning/concept-workspace?view=azureml-api-2).
2. this exercise assumes that you've already downloaded the [Credit Card Fraud Detection](https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud) dataset from kaggle, and stored this in 'Assets -> Data' inside your Azure account under the name, 'creditcard_fraud'.**

### Summary

This proof-of-concept demonstrates how machine learning may be effectively leveraged to identify fraudulent credit card transactions when it is needed most:  at the time of the transaction. 

This document outlines a process to implement a production ready solution.

Success depends on leveraging the organization’s unique capabilities and data assets, with structured opportunities for stakeholders across risk, operations, and customer experience to contribute to ongoing model improvement.

#### The Objective

Fraudulent transactions are rare - in our proof-of-concept dataset, only 0.17% are proven to be fraudulent. This type of problem is classified as anomaly detection and has two relevant metrics: *Recall* and *Precision*. These terms will be used throughout the process.

- *Recall* is a _"Of all actual fraud cases, how many did we catch?"_
    - Low percentages mean less fraudulent transactions are flagged.
    - Therefore, lower recall implies *higher operation costs* incurred in dealing with fraud after the settlement of the transaction.
- *Precision* is _“Of all transactions we flagged as fraud, how many were actually fraud?”_
    - Low percentages mean normal transactions will be flagged as fraud.
    - This *negatively affects the perception of value by customers as this negatively impacts their usage of our service*.

#### The Process

Using historical transaction data, we train an anomaly detection model to classify transactions as normal or fraudulent. 

This class of machine learning - anomaly detection - is a complex but achievable problem with several open source models available to use.

During the process, our goal is to balance the two measures of Recall and Precision to rates aligned your organization's strategy.

Model Training is configuring an algorithm with customized values for a large amount of variables.

A simplified analogy is a vehicle that stores driver seat settings. Initially, the driver adjusts the seat manually based on comfort. Over time, the system remembers these adjustments and applies them automatically. Similarly, a machine learning model learns patterns from historical data and uses those patterns to make future predictions, with periodic updates as new information becomes available.

## Workflow
### Diagram
![Workflow Process for ML Training - Prep, Model Training, Selection, and Monitoring with stakeholder feedback](img/ML_Fraud_Workflow_Feedback.png)

### 1. Data Preparation & Training Setup

| Azure ML Component | Role in the Process |
| ------------------ | ------------------- |
| 1 a) Dataset (Credit Card Data) | Provides historical transaction data used to train and evaluate the model. |
| 1 b) Jupyter Notebook (Azure ML / VS Code) | Used to explore data, document the workflow, and prepare features for modeling. |
| 1 c) Data Processing (Prepare the Data) | Transforms raw transaction data into structured inputs suitable for machine learning. |
| 1 d) Train/Test Split | Separates data into training and validation sets to ensure unbiased model evaluation. |

#### Preliminary Setup Steps

In [2]:
# Step 1: Import Packages and Connect to your Azure Workspace
from azureml.core import Workspace, Dataset          # see https://pypi.org/project/azureml-core/
import pandas as pd                                  # see https://pandas.pydata.org/docs/
from sklearn.ensemble import IsolationForest         # see https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.IsolationForest.html
from sklearn.metrics import classification_report    # see https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html
from azureml.core.model import Model                 # see https://docs.microsoft.com/en-us/python/api/azureml-core/azureml.core.model?view=azure-ml-py 
from sklearn.model_selection import train_test_split # see https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html


#### 1a) Dataset (Credit Card Data)

To ensure accurate training, the dataset *must* be kept in a *known location* where the state of the data is tracked and known. We recommend using the features of Kaggle.com or AzureML for this purpose.

#### 1b) Jupyter Notebook (Python / Azure ML / VS Code)

This allows for analysis and exploration of data in a tracked and reproducible way. Given the complexity, usage of such tools is highly recommended for this type of work. This notebook is a Jupyter Notebook.

Python is favored for this activity as a large community of data scientists and machine learning experts have creating free and open source modules for all to use.

In [ ]:
import pandas as pd
from pathlib import Path

# =========================
# LOCAL MODE (default)
# =========================
path = Path("data/creditcard.csv")

if path.exists():
    print("Loading dataset from local file...")
    df = pd.read_csv(path)

# =========================
# AZURE MODE (fallback)
# =========================
else:
    print("Local file not found. Attempting Azure ML load...")
    
    from azureml.core import Workspace, Dataset
    
    ws = Workspace.from_config()  # assumes config.json is present
    dataset = Dataset.get_by_name(ws, name='creditcard_fraud')
    df = dataset.to_pandas_dataframe()

# Preview data
print("Dataset shape:", df.shape)
df.head()

Loading dataset from local file...
Dataset shape: (284807, 31)


,Time,V1,V2,V3,V4,V5,V6,V7,V8,V9,...,V21,V22,V23,V24,V25,V26,V27,V28,Amount,Class
0,0.0,-1.359807,-0.072781,2.536347,1.378155,-0.338321,0.462388,0.239599,0.098698,0.363787,...,-0.018307,0.277838,-0.110474,0.066928,0.128539,-0.189115,0.133558,-0.021053,149.62,0
1,0.0,1.191857,0.266151,0.166480,0.448154,0.060018,-0.082361,-0.078803,0.085102,-0.255425,...,-0.225775,-0.638672,0.101288,-0.339846,0.167170,0.125895,-0.008983,0.014724,2.69,0
2,1.0,-1.358354,-1.340163,1.773209,0.379780,-0.503198,1.800499,0.791461,0.247676,-1.514654,...,0.247998,0.771679,0.909412,-0.689281,-0.327642,-0.139097,-0.055353,-0.059752,378.66,0
3,1.0,-0.966272,-0.185226,1.792993,-0.863291,-0.010309,1.247203,0.237609,0.377436,-1.387024,...,-0.108300,0.005274,-0.190321,-1.175575,0.647376,-0.221929,0.062723,0.061458,123.50,0
4,2.0,-1.158233,0.877737,1.548718,0.403034,-0.407193,0.095921,0.592941,-0.270533,0.817739,...,-0.009431,0.798278,-0.137458,0.141267,-0.206010,0.502292,0.219422,0.215153,69.99,0


#### 1c) Data Processing (Prepare the data)

Sometimes data must be structured for suitability for model training and analysis. For our proof of concept, we're going to normalize the distribution of the transaction $ amount column, which helps the model treat transaction amounts on the same scale as the other normalized features.

We also drop the `Time` column since it doesn't contribute meaningfully to anomaly detection in this context.

In [22]:
df['Amount'] = (df['Amount'] - df['Amount'].mean()) / df['Amount'].std()
X = df.drop(columns=['Class', 'Time'])
y = df['Class']


#### 1d) Train/Test Split

The concept of Train/Test splitting is fundamental to Machine Learning and Model Training. 

- Training data is *only* used in the model training.
- Testing data is *only* used to evaluate the training results of Model Training.

If Testing data were to be used in training data, it would compromise the integrity of the evaluation process.

In [26]:
# Split into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(
    X, 
    y, 
    test_size=0.2,      # 20% test, 80% train
    random_state=42,    # ensures reproducibility
    stratify=y          # IMPORTANT for imbalanced data
)

print("Training samples:", X_train.shape[0])
print("Testing samples:", X_test.shape[0])

Training samples: 227845
Testing samples: 56962


*Additional Information* Note the use of stratify - This ensures the fraud is equally represented among the testing and training datasets. It is concepts like this that require a holistic understanding of data of which data scientists have an understanding of.

### 2. Model Training & Evaluation

| Azure ML Component or Processing Step | Role in the Process |
| ----------------------------------------------------------- | ------------------------------------------------------------------------------------ |
| 2a) Compute Instance / Compute Cluster | Executes model training and evaluation processes |
| 2b) Experiment | Tracks multiple model runs and configurations for comparison |
| 2c) Models (Logistic Regression, Isolation Forest, Autoencoder) | Different algorithms used to detect fraudulent patterns in transaction data |
| 2d) Evaluation Metrics (Precision, Recall, F1-Score) | Measures model performance, with emphasis on detecting rare fraud cases effectively |

#### 2a) Compute Instance / Compute Cluster

Azure ML compute resources are used to execute model training and evaluation tasks. These environments provide the processing power required to efficiently train models on large transaction datasets and support iterative experimentation.

As always, there is a balance between cost, time, and money that must be balanced. It is recommended to allow for budgeted access to these resources so these risk assessments are performed by the experts.

#### 2b) Experiment

Each model run is recorded as part of an experiment, enabling comparison across different algorithms, configurations, and parameters. This ensures that model selection is based on measurable performance rather than assumptions.

The Azure ML environment enables the automation of such experiments by providing a drag and drop interface built to effectively perform such analysis.

#### 2c) Models (Logistic Regression, Isolation Forest, Autoencoder)

Multiple modeling approaches are evaluated to capture different types of fraud patterns:

- *Logistic Regression* identifies linear relationships between transaction features and fraud likelihood
- *Isolation* Forest detects anomalies by isolating unusual patterns in the data
- *Autoencoder* learns typical transaction behavior and flags deviations as potential fraud

Evaluating multiple models ensures that the selected approach aligns with both the statistical characteristics of the data and the organization’s risk tolerance.


In [27]:
# An example of one training instance with two different items being configured.
# In practice, this is ran multiple times to create a model ready for organizational
# review and potential deployment to our Point of Sale solutions.

# Initialize model
model = IsolationForest(contamination=0.0017, random_state=42)

# Train on training data only
model.fit(X_train)

# Predict on test data
y_pred = model.predict(X_test)

# Convert output: -1 → 1 (fraud), 1 → 0 (normal)
y_pred = [1 if x == -1 else 0 for x in y_pred]

#### 2d) Evaluation Metrics (Precision, Recall, F1-Score)

The table below is a summary of metrics that are calculated from the [confusion matrix](https://en.wikipedia.org/wiki/Confusion_matrix). It shows how well our model identified normal and fraudulent transactions:

| Metric       | What It Means |
|--------------|-----------------|
| **Recall**    | Of all actual fraud cases, how many did we catch? |
| **Precision** | Of all transactions we flagged as fraud, how many were actually fraud? |
| **F1-Score**  | A balance between precision and recall — like a combined performance score. |
| **Support**   | The number of examples in each group (normal or fraud) in the real data. |

In [ ]:
#Use the testing data, which we had previously set aside for evaluation, to get a score for the performance of the train

print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     56864
           1       0.32      0.33      0.32        98

    accuracy                           1.00     56962
   macro avg       0.66      0.66      0.66     56962
weighted avg       1.00      1.00      1.00     56962



##### Proof of Concept Results Summary and Demonstration of Precision and Recall

| Class | Description           | Precision | Recall | F1-Score | Support |
|-------|------------------------|-----------|--------|----------|---------|
| `0`   | Normal transactions    | **1.00**  | **1.00** | **1.00**   | 56,864 |
| `1`   | Fraudulent transactions| **0.32**  | **0.33** | **0.32**   | 98     |

##### Interpretation (In Simple Terms)

- The model is **excellent at recognizing normal transactions** — it almost never makes a mistake with those.
- However, it **struggles to correctly catch fraud**:
  - Precision: When it says a transaction is fraud, it’s **only right 32% of the time**.
  - Recall: It **only finds 33% of the real fraud cases** — it misses most of them.

##### Overall Accuracy

- The model is **99.9% accurate**, but this is misleading.
- Because **fraud cases are very rare**, the model can look “perfect” just by saying everything is normal.
- That’s why we look at **precision**, **recall**, and **F1-score** for a fuller picture.

##### Is it ready for production?

- No - there are far too many false positives evident in these results. Additional analysis is needed.

### 3. Decision & Scoring Layer

| Azure ML Component or Process Step | Role in the Process |
| ------------------ | ------------------ |
| 3a) Model Selection | Identifies the model that best balances fraud detection with operational impact |
| 3b) Prediction / Scoring | Applies the selected model to assign fraud risk scores to transactions |
| 3c) Business Decision Layer | Determines actions such as approving, flagging, or investigating transactions |

#### 3a) Model Selection 

In this phase, the selected model is applied to real-world transactions, and its outputs are translated into actionable business decisions. The focus is not only on model performance, but on how predictions are used to manage fraud risk while maintaining a positive customer experience.

The model chosen from the evaluation phase is selected based on its ability to balance fraud detection (recall) with the operational impact of false positives. This decision reflects the organization’s risk tolerance, recognizing that overly aggressive detection may disrupt legitimate transactions, while overly lenient detection may allow fraud to occur.

#### 3b) Prediction / Scoring

The selected model is used to assign a fraud risk score to each transaction. Rather than producing a simple “fraud” or “not fraud” output, the model generates a probability or score indicating the likelihood of fraud. This enables more flexible and nuanced decision-making.

#### 3c) Business Decision Layer

Model outputs are translated into operational actions based on predefined thresholds and business rules:

- *Approve* - Low-risk transactions are processed without interruption
- *Flag* - Medium-risk transactions may trigger alerts or additional verification
- *Investigate* - High-risk transactions are escalated for manual review

These decisions are designed to balance fraud prevention with customer experience and operational efficiency. Thresholds can be adjusted over time to reflect changing business priorities and evolving fraud patterns.

### 4. Continuous Feedback & Improvement

*Continuous feedback and refinement are essential to maintaining model effectiveness, as fraud patterns evolve and business priorities shift over time.*

| Azure ML Component | Role in the Process |
| ------------------ | ------------------- |
| Monitoring & Logging | Tracks ongoing model performance, including false positives and missed fraud |
| Stakeholder Input (Risk, Operations, Customer Experience) | Provides operational feedback on model effectiveness and business impact |
| Retraining Pipeline (Future State) | Incorporates new data and feedback to refine and improve model performance over time |

Fraud detection is not a one-time implementation but an ongoing process. This phase ensures that the model remains effective over time by incorporating performance monitoring, stakeholder input, and iterative refinement.

#### 4a) Monitoring & Logging

Once the model is applied to real-world transactions, its performance is continuously monitored. Key indicators include false positives (legitimate transactions incorrectly flagged) and missed fraud cases. Tracking these metrics over time helps identify performance degradation, emerging fraud patterns, and potential model drift.

#### 4b) Stakeholder Input (Risk, Operations, Customer Experience)

Operational teams provide critical feedback on how the model performs in practice:

- Risk & Compliance Teams assess whether fraud detection levels meet organizational and regulatory expectations
- Operations Teams evaluate the impact on investigation workload and processing efficiency
- Customer Experience Teams monitor the effect of false positives on customer satisfaction

This feedback is structured and translated into measurable improvements rather than directly altering the model.

#### 4c) Retraining Pipeline (Future State)

Over time, new transaction outcomes (e.g., confirmed fraud or false alarms) are incorporated into the dataset. This enriched data is used to retrain the model, improving its ability to detect evolving fraud patterns.

Retraining may include:

- Updating model parameters
- Incorporating new features
- Adjusting decision thresholds

This iterative process ensures that the model adapts to changing conditions and continues to align with business objectives.